tests/test_options.py

```diff
@@ -212,3 +264,365 @@ def test_as_dict():
+    assert Options()['model.criterion'] == 'new value'
+
+
+def test_getattr():
+    reset_options_instance()
+    source = {'abc': 123}
+    Options(source, run_parser=False)
+    assert Options().options == source
+    assert Options().abc == 123
+
+
+def test_get_exist_value():
+    reset_options_instance()
+    source = {'abc': 123}
+    Options(source, run_parser=False)
+    assert Options().options == source
+    value = Options().get('abc', 'default value')
+    assert value == 123
+
+
+def test_get_default_value():
+    reset_options_instance()
+    source = {'abc': 123}
+    Options(source, run_parser=False)
+    assert Options().options == source
+    value = Options().get('cba', 'default value')
+    assert value == 'default value'
+
+
+def test_has_key_true():
+    reset_options_instance()
+    source = {'abc': 123}
+    Options(source, run_parser=False)
+    assert Options().options == source
+    assert Options().has_key('abc')
+
+
+def test_has_key_false():
+    reset_options_instance()
+    source = {'abc': 123}
+    Options(source, run_parser=False)
+    assert Options().options == source
+    assert not Options().has_key('cba')
+
+
+def test_keys():
+    reset_options_instance()
+    source = {
+        'model': 'mymodel',
+        'dataset': 'mydataset'
+    }
+    Options(source, run_parser=False)
+    assert Options().options == source
+    assert sorted(Options().keys()) == sorted(['model', 'dataset'])
+
+
+def test_values():
+    reset_options_instance()
+    source = {
+        'model': 'mymodel',
+        'dataset': 'mydataset'
+    }
+    Options(source, run_parser=False)
+    assert Options().options == source
+    assert sorted(Options().values()) == sorted(['mymodel', 'mydataset'])
+
+
+def test_items():
+    reset_options_instance()
+    source = {'model': 'mymodel'}
+    Options(source, run_parser=False)
+    assert Options().options == source
+    for key, value in Options().items():
+        assert key == 'model'
+        assert value == 'mymodel'
+
+
+def test_lock():
+    reset_options_instance()
+    source = {
+        'dataset': 123,
+        'model': {
+            'criterion': 'I am a criterion',
+            'network': 'I am a network',
+        },
+    }
+    Options(source, run_parser=False)
+    assert Options().options == source
+    Options().unlock()
+    assert not Options().options.islocked()
+    assert not Options().options['model'].islocked()
+    Options().lock()
+    assert Options().options.islocked()
+    assert Options().options['model'].islocked()
+
+
+def test_unlock():
+    reset_options_instance()
+    source = {
+        'dataset': 123,
+        'model': {
+            'criterion': 'I am a criterion',
+            'network': 'I am a network',
+        },
+    }
+    Options(source, run_parser=False)
+    assert Options().options == source
+    Options().lock()
+    assert Options().options.islocked()
+    assert Options().options['model'].islocked()
+    Options().unlock()
```

Comment:
This line throws a lot of messages into stdout. We could capture them by replacing the stream of sys.stdout with a string stream and checking if the messages were printed; it could be a safety check test :)